FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RefreshPrivateKeysFromEkmViewModel.kt

```diff
@@ -0,0 +1,233 @@
+    var passphrase: Passphrase? = null
+
+    for (fetchedPgpKeyDetails in fetchedPgpKeyDetailsList) {
+      if (!fetchedPgpKeyDetails.usableForEncryption) {
+        //ask Tom
+        //we skip a key that is not usable for encryption
+        continue
+      }
+
+      val existingPgpKeyDetails =
+        existingPgpKeyDetailsList.firstOrNull { it.fingerprint == fetchedPgpKeyDetails.fingerprint }
+
+      if (existingPgpKeyDetails != null) {
+        if (fetchedPgpKeyDetails.isNewerThan(existingPgpKeyDetails)) {
+          if (passphrase == null) {
+            passphrase = getUsablePassphraseFromCache()
+          }
+
+          val existingKeyEntity = existingKeyEntities.first {
+            it.fingerprint == existingPgpKeyDetails.fingerprint
+          }
+
+          val safeVersionOfPrvKey = protectAndEncryptInternally(passphrase, fetchedPgpKeyDetails)
+          roomDatabase.keysDao().updateSuspend(
+            existingKeyEntity.copy(
+              privateKey = safeVersionOfPrvKey,
+              publicKey = fetchedPgpKeyDetails.publicKey.toByteArray()
+            )
+          )
+        }
+      } else {
+        if (passphrase == null) {
+          passphrase = getUsablePassphraseFromCache()
+        }
+
+        val safeVersionOfPrvKey = protectAndEncryptInternally(passphrase, fetchedPgpKeyDetails)
+        val keyEntity = fetchedPgpKeyDetails.toKeyEntity(activeAccount).copy(
+          privateKey = safeVersionOfPrvKey,
+          storedPassphrase = null
+        )
+        roomDatabase.keysDao().insertSuspend(keyEntity)
+      }
+    }
+  }
+
+  /**
+   * We receive a private key from EKM in decrypted format. Before saving it to the local database
+   * we have to protect it with a provided [Passphrase] and encrypt it with [AndroidKeyStore]
+   */
+  private suspend fun protectAndEncryptInternally(
+    passphrase: Passphrase,
+    pgpKeyDetails: PgpKeyDetails
+  ): ByteArray = withContext(Dispatchers.IO) {
+    val protectedPrvKey = PgpKey.encryptKeySuspend(
+      armored = requireNotNull(pgpKeyDetails.privateKey),
+      passphrase = passphrase
+    )
+    val encryptedPrvKeyInternally = KeyStoreCryptoManager.encryptSuspend(protectedPrvKey)
+    return@withContext encryptedPrvKeyInternally.toByteArray()
+  }
+
+  /**
+   * Get the first non-null pass phrase for any of the existing keys already in storage
+   * or throw [EmptyPassphraseException] if it is not exist.
+   */
+  private fun getUsablePassphraseFromCache(): Passphrase {
+    val context: Context = getApplication()
+    val keysStorage = KeysStorageImpl.getInstance(context)
+    val fingerprintOfKeyWithNotEmptyPassphrase = keysStorage.getRawKeys()
+      .map { it.fingerprint }
+      .firstOrNull {
+        keysStorage.getPassphraseByFingerprint(it)?.isEmpty == false
+      } ?: throw EmptyPassphraseException(
+      fingerprints = keysStorage.getFingerprintsWithEmptyPassphrase(),
+      message = context.getString(R.string.empty_pass_phrase)
+    )
```

Comment:
I'm assuming that some code catches this and displays a modal to user? And then the whole operation is restarted? Is EKM API also re-pulled?

Reply:
You are right. I'm posting places.

> I'm assuming that some code catches this and displays a modal to user? 

https://github.com/FlowCrypt/flowcrypt-android/blob/61f73f5c4d8821f250363ac4dcd738f8424f1e55/FlowCrypt/src/main/java/com/flowcrypt/email/ui/activity/MainActivity.kt#L437

> And then the whole operation is restarted? Is EKM API also re-pulled?

https://github.com/FlowCrypt/flowcrypt-android/blob/61f73f5c4d8821f250363ac4dcd738f8424f1e55/FlowCrypt/src/main/java/com/flowcrypt/email/ui/activity/MainActivity.kt#L464